[RISCV] add more generic macrofusions

danielhb opened the pull request:
These are some macrofusions that are used internally in Ventana in an yet not upstreamed processor. Figured it would be good to contribute them ahead of the processor to allow the community to also use them in their own processors, while also alleaviting our own downstream upkeep.

The macrofusions being added are, considering load = lb,lh,lw,ld,lbu,lhu,lwu:

- bfext (slli+srli)
- auipc+load
- lui+load
- add(.uw)+load 
- addi+load
- shXadd(.uw)+load, where X=1,2,3

danielhb committed:
[RISCV] add more generic macrofusions

These are some macrofusions that are used internally in Ventana in an
yet not upstreamed processor. Figured it would be good to contribute
them ahead of the processor to allow the community to also use them in
their own processors, while also alleaviting our own downstream upkeep.

The macrofusions being added are:

- add+lw
- addi+ld, addi+lw
- adduw+lw
- auipc+ld
- bfext (slli+srli)
- lui+ld
- shXadd+load, where X=1,2,3 and load=lb,lh,lw,ld
- shXadduw+load, where X=1,2,3 and load=lb,lh,lw,ld

danielhb commented on llvm/lib/Target/RISCV/RISCVMacroFusion.td:
```diff
@@ -91,3 +91,94 @@ def TuneLDADDFusion
+
+// Fuse add with lw:
+//   add rd, rs1, rs2
+//   lw rd, 0(rd)
```
Oh, I believe I forgot add+lh/add+lb because they barely appear in the benchmarks I've run. The bulk of fusions are add+ld and add+lw and LLVM already has add+ld.

I'll add add+lh and add+lb to complete the add + load macrofusion support. Should I add an aggregate add+ld/lw/lh/lb macrofusion like I did with shNadd?

topperc replied:
Sorry I didn't realized we had TuneLDADDFusion already. Looks like it is only used by ventana-veyron-v1. Is that the only add+load fusion implemented by ventana-veyron-v1?

danielhb replied:
Just double checked the design docs:

- veyron-v1 implements add+ld and add+lw only. And it also implements auipc+ld and lui+ld. All of them added by this PR but none of them actually added in the processor description. I'll add them;

- for the rest of macrofusions being added, and already answering the questions below, 'load'  consists of lb/lh/lw/ld/lbu/lhu/lwu. Aside from the ld and lw I don't see expressive macrofusions numbers of add (and shNadd) with lb/lh/lbu/lhu/lwu, but I think it's clearer to just add everything.  Also, the numbers I'm seeing might not be representative of what every other processor will see and perhaps these macrofusions are more common than I might realize ...

danielhb committed:
Add extra load macrofusion cases and other fixes

- add missing macrofusions in veyron-v1 processor def;

- fix ADD_UW/add.uw comment;

- add ADD + lb/lh/lbu/lhu/lwu macrofusions;

- add shXADD + lbu/lhu/lwu macrofusions;

- add shXADD_UW + lbu/lhu/lwu macrofusions.

danielhb commented on llvm/lib/Target/RISCV/RISCVMacroFusion.td:
```diff
@@ -91,3 +91,106 @@ def TuneLDADDFusion
+
+// Fuse AUIPC followed by LD:
+//   auipc rd, imm20
+//   ld rd, imm12(rd)
```
I pulled this one and LUI+LD from the GCC support we have upstream and didn't really check in the docs. After reading them, no, it's not only 64 bit loads for neither. It's lb/lh/lw/ld/lbu/lhu/lwu (what we're calling 'load').

This is also true for add.uw + lw (i.e it's add.uw + load)  and addi + ld/lw (it's addi + load). 

And I missed AUIPC + ADDI/ADDIW and LUI+ADDI/ADDIW, something that veyron-v1 doesn't support but the new engine does. I'll add them up.

danielhb commented on llvm/lib/Target/RISCV/RISCVMacroFusion.td:
```diff
@@ -91,3 +91,106 @@ def TuneLDADDFusion
+                 "Enable SLLI+SRLI (bitfield extract) macrofusion",
+                 CheckAll<[
+                   CheckOpcode<[SLLI]>,
+                   CheckIsImmOperand<2>,
```
It's not. I'll remove it

danielhb committed:
RISCVMacroFusion.td: review changes

- add TuneAUIPCADDIWFusion;

- turn TuneAUIPCLDFusion (auipc+ld) into TuneAUIPCLoadFusion
  (auipc + lb/lh/lw/ld/lbu/lhu/lwu);

- turn TuneLUILDFusion (lui+ld) into TuneLUILoadFusion
  (lui + lb/lh/lw/ld/lbu/lhu/lwu);

- turn TuneADD_UWLWFusion (add.uw+lw) into TuneADD_UWLoadFusion
  (add.uw + lb/lh/lw/ld/lbu/lhu/lwu);

- remove TuneADDILWFusion. Turn TuneADDILDFusion into TuneADDILoadFusion
  (addi + lb/lh/lw/ld/lbu/lhu/lwu);

- remove the immediate check from TuneADDLoadFusion;

- remove the immediate check from TuneBFExtFusion: it was a copy/paste
  reminiscent from the existing slli+srli fusions;

- renames: TuneSHXADD_UWLoadFusion and "Enable SH(1|2|3)ADD_UW ..."

wangpc-pp reviewed (commented):
I wonder if we can merge some fusions into one? For example, `add+lw` and `add.uw+lw`, I don't see why we should make them standalone fusions.

danielhb commented on llvm/lib/Target/RISCV/RISCVMacroFusion.td:
```diff
@@ -28,6 +28,15 @@ def TuneAUIPCADDIFusion
+  : SimpleFusion<"auipc-addiw-fusion", "HasAUIPCADDIWFusion",
+                 "Enable AUIPC+ADDIW macrofusion",
+                 CheckOpcode<[AUIPC]>,
+                 CheckOpcode<[ADDIW]>>;
```
This was added because the design happens to support it, but we can remove it. In fact I struggle to see this one in my spec runs, although auipc + addi is very popular

danielhb commented on llvm/lib/Target/RISCV/RISCVMacroFusion.td:
```diff
@@ -91,3 +100,87 @@ def TuneLDADDFusion
+
+// Fuse add followed by a load (lb, lh, lw, ld, lbu, lhu, lwu):
+//   add rd, rs1, rs2
+//   load rd, 0(rd)
```
It doesn't. I forgot to update the comment after removing the immediate check. I'll fix it

danielhb committed:
Remove and consolidate macrofusions

- remove auipc+addiw since it rarely/doesn't happen;
- remove add+lw. veyron-v1 now uses add+load;
- merge add+load and add.uw+load into a single fusion;
- merge shXadd+load and shXadd.uw+load into a single fusion;
- fix immediate 0 comment in AddLoad fusion.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVMacroFusion.td:
```diff
@@ -91,3 +91,57 @@ def TuneLDADDFusion
+  : SimpleFusion<"addi-load-fusion", "HasADDILoadFusion",
+                 "Enable ADDI + load macrofusion",
+                 CheckOpcode<[ADDI]>,
+                 CheckOpcode<[LB, LH, LW, LD, LBU, LHU, LWU]>>;
```
We can use `defvar` here to remove redundancies.
```
defvar Load = [LB, LH, LW, LD, LBU, LHU, LWU];
```

danielhb committed:
Add 'Load' var to reduce repetition of load opcodes

wangpc-pp reviewed (approved):
LGTM. Cheers!

topperc reviewed (approved):
LGTM
